modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/DataApplicationSnapshotAcceptanceManager.scala

```diff
@@ -103,39 +152,43 @@ object DataApplicationSnapshotAcceptanceManager {
                 .ifF(updates.toList, List.empty[Signed[DataUpdate]])
             }
             .getOrElse(List.empty[Signed[DataUpdate]].pure[F])
-            .flatMap { updates =>
-              service.getCalculatedState
-                .flatMap(expectCalculatedStateOrdinal(lastOrdinal))
-                .map(DataState(lastState, _))
-                .flatMap(service.combine(_, updates))
-                .flatTap(state => service.setCalculatedState(currentOrdinal, state.calculated))
-                .flatMap(state => service.serializeState(state.onChain))
-            }
-            .map(_.some)
-            .handleErrorWith(err =>
-              logger
-                .error(err)(
-                  "Unhandled exception during combine and serialize data application, fallback to last data application"
-                )
-                .as(maybeLastDataApplication)
-            )
         }
 
-        calculatedState <- service.getCalculatedState
-          .flatMap(expectCalculatedStateOrdinal(currentOrdinal))
-
-        serializedDataBlocks <- dataBlocks.traverse(service.serializeBlock)
-
-        calculatedStateHash <- service.hashCalculatedState(calculatedState)
-
-        result = maybeNewDataState.map { state =>
-          DataApplicationAcceptanceResult(
-            DataApplicationPart(state, serializedDataBlocks, calculatedStateHash),
-            calculatedState
-          )
-
+        newDataState <- OptionT.liftF(
+          service.combine(DataState(lastOnChainState, lastCalculatedState), updates)
+        )
+
+        serializedOnChainState <- OptionT.liftF(
+          service.serializeState(newDataState.onChain)
+        )
+
+        serializedBlocks <- OptionT.liftF(
+          dataBlocks.traverse(service.serializeBlock)
+        )
+
+        calculatedStateProof <- OptionT.liftF(
+          service.hashCalculatedState(newDataState.calculated)
+        )
+
+      } yield
+        DataApplicationAcceptanceResult(
+          DataApplicationPart(serializedOnChainState, serializedBlocks, calculatedStateProof),
+          newDataState.calculated
+        )
+
+      newDataState.value.handleErrorWith { err =>
+        logger.error(err)("Unhandled exception during calculating new data application state, fallback to last data application").flatMap {
+          _ =>
+            service.getCalculatedState.map { lastCalculatedState =>
```

Comment:
>Shouldn't you just return the constructed DataApplicationAcceptanceResult

I am doing that. The code you are referring to is in `handleErrorWith` so in case of any error we make a fallback to the old `DataApplicationAcceptanceResult`

Reply:
Ok it's a fallback when the flow fails, so I guess it's ok.

Reply:
Yeah I noticed, makes sense.

Reply:
> > Shouldn't you just return the constructed DataApplicationAcceptanceResult
> 
> I am doing that. The code you are referring to is in `handleErrorWith` so in case of any error we make a fallback to the old `DataApplicationAcceptanceResult`

Yeah I noticed, makes sense.